### DISCUSSION TITLE
Enh ideas: export curves of voltage, frequency, gas consumption/kw over time

### DISCUSSION THREAD

**Discussion by billpribyl** [jgyates/genmon#1119] [Ideas]
I know that someone (probably not me) could write a custom app that would use the web API to call cmd/status_json and cmd/maint_json to periodically grab the status and maintenance data, and record it in a database for analysis.  (I suppose it might also be possible to grab the data via MQTT, but I have been unable to get MQTT to install here so haven't been able to play with that.)

The goal would be to chart and/or analyze the data over time, to get a measure of performance and possible warning signs if they go off nominal. At the moment I'm interested in:

1. Voltage over time
2. Frequency over time
3. Rate of gas consumption per kw generated over time

During the recent 6-day outage at my house, I noticed a bit of voltage and frequency drift when running under load, but I don't really have a way to guage the magnitude or prevalence of these variations. Maybe they were within the "normal" range, maybe not -- I can't really tell unless I can look at a historical curve. (And I'm not entirely sure what "normal" is for my generator).

In addition to understanding whether the generator is working normally, #3 would be super helpful in planning appliance use during an outage.

Has this already been done?

Thanks!

**Comment by jgyates** [comment]
Hi @billpribyl 

The supported way to do this is via MQTT. You can export almost all of the data that genmon displays via MQTT. Some people use [Home Assistant](https://github.com/jgyates/genmon/issues/155), other use [Grafana](https://github.com/jgyates/genmon/issues/673). Both of these approaches use MQTT to export the data. I personally do this:

I have an MQTT server (MQTT mosquito broker) that does nothing but receive data from genmon (and some other things in my house). I use Node Red to take the data from the MQTT broker and put it in an InfluxDB data base. Then I created a Grafana dashboard to display the data.

Grafana dashboard for genmon data:

<img width="1692" alt="Screenshot 2024-05-23 at 5 37 48 PM" src="https://github.com/jgyates/genmon/assets/4513286/648f81dc-80ad-461b-927b-1634c9d3cca1">

I have the MQTT broker, Node Red, InfluxDb and Grafana all running in docker cotainers on a server. As an experiment I set this up twice, once one a rasbperry pi with a 1TB SD card and once on a NAS server.

I have you questions on MQTT setup or any of this you can post questions here.

**Comment by billpribyl** [comment]
Thanks, it makes sense. I tried to install Mosquito MQTT from the Home Assistant Add-on Store on my Pi 3B+ Rev 1.3 running HA 2024.5.x on HAOS 12.3. (This is not the same Pi on which I'm running genmon). It claimed to install, but it wouldn't run. It would spin and spin and eventually say something like "failed, check supervisor" which was a bit mysterious to me.  I did look at the supervisor log file and there was nothing there which I found helpful. When I tried to reboot HA, it took a looooong time...maybe 30 minutes...and I eventually had to uninstall Mosquito broker to make my HA usable again. The HA Pi has a 32G SD card with quite a lot of free space. I suppose my Pi might be underpowered, but Mosquito's web site says it is "lightweight and is suitable for use on all devices from low power single board computers to full servers," so I don't really know.

**Comment by jgyates** [comment]
I set it up on a spare pi using [IOTstack](https://github.com/SensorsIot/IOTstack). Here is the article I used to setup the pi with IOTStack: https://learnembeddedsystems.co.uk/easy-raspberry-pi-iot-server

Here is the IOTStack wiki: https://sensorsiot.github.io/IOTstack/

The IOTStack project will add docker and install all the needed software as docker images.

You can setup the influxDB to have a maximum retention period, but once you set the max retention all the data will be reset. IOTstack can use influxDB or influxDB2. influxDB is easier, but influxDB3 allows more flexibility in displaying queries over time.

**Comment by billpribyl** [comment]
I have configured at MQTT broker, and genmon is actively sending messages to it. I am able to view the messages using the MQTTX toolbox. I'm seeing MQTT messages with topics such as Utility Voltage, Battery Voltage, Board Temp, Run Hours in the last 30 days, and Ambient Temperature Sensor.  

But I'm not seeing some other data that I would hope to see, such as WiFi signal strength, Fuel Consumption in last 30 days, Total Power Log Fuel Consumption. I have not tested under load, but during the latest scheduled "quiet mode" exercise I would have also expected to see RPM, Output Voltage, Frequency, Current L1, Current L2, Power Output...but I am not seeing any of those come through MQTTX either.

Am I doing something wrong?

Thanks

**Reply by jgyates** [reply]
@billpribyl 

The MQTT protocol is designed to optimize network traffic so you can configure sensor endpoints like the genmon MQTT add on to only send data when it changes and optionally set a timeout value that the add on will send all of the data, regardless if it has changed. I would take a look at the genmon MQTT add on setting, specifically the "Flush Interval" setting. The data you are reporting to see (Utility Voltage, Battery Voltage, Temp, etc) all change frequency so I am guessing you have not set Flush Interval to a value that meets your needs.

Let me know if that does not solve your issue.
